### PR TITLE
Fix ambiguous data issues

### PR DESCRIPTION
**Pull request by oxidase**
Added disambiguations for edge weights and names by taking minimal edge weight and minimal name id for edges with equal source and target nodes.

Solves  #2617

### CODE REVIEW

File: src/extractor/extraction_containers.cpp [hunk 1]
```diff
@@ -453,11 +453,13 @@ void ExtractionContainers::PrepareEdges(lua_State *segment_state)
                 all_edges_list[i].result.weight < min_forward_weight)
             {
                 min_forward_idx = i;
+                min_forward_weight = all_edges_list[i].result.weight;
```
Comment (TheMarex) [inline src/extractor/extraction_containers.cpp]: This would mean the de-duplication code has been broken for a while?

File: include/extractor/internal_extractor_edge.hpp [hunk 2]
```diff
@@ -142,14 +142,16 @@ struct InternalExtractorEdge
...
 {
     using value_type = InternalExtractorEdge;
     bool operator()(const InternalExtractorEdge &lhs, const InternalExtractorEdge &rhs) const
     {
         return (lhs.result.source < rhs.result.source) ||
                ((lhs.result.source == rhs.result.source) &&
-                (lhs.result.target < rhs.result.target));
+                (lhs.result.target < rhs.result.target ||
+                 (lhs.result.target == rhs.result.target &&
+                  lhs.result.name_id < rhs.result.name_id)));
```
Comment (TheMarex) [inline include/extractor/internal_extractor_edge.hpp]: Usually having a name is better than not having a name so inverting this is probably more what we want.

**Comment by TheMarex** [comment]
Good catch. This must have been a very old bug, at least going back a year.

File: src/extractor/extraction_containers.cpp (hunk 1 above)
Comment (oxidase) [inline src/extractor/extraction_containers.cpp]: Yes, always the last edge was used. Because the order of edges was (source, target),  edge weights were "randomized" by scheduler.

File: include/extractor/internal_extractor_edge.hpp (hunk 2 above)
Comment (oxidase) [inline include/extractor/internal_extractor_edge.hpp]: Ok, i will revert the order.

**Comment by systemed** [comment]
Very interesting. I've occasionally found issues similar to #610 where OSRM appears to pick the less favourable way between two nodes - for example, at http://cycle.travel/map?from=51.4691,-0.2196&to=51.4707,-0.2216, where the middle section is seemingly following a highway=pedestrian area even though the road has a better weight. Hopefully this fixes it.

**Comment by oxidase** [comment]
@systemed looks like it can be solved by PR.

**Comment by emiltin** [comment]
Would be really nice if this could solve #610 - which is over 3 years old :-)

File: include/extractor/internal_extractor_edge.hpp [hunk 3]
```diff
@@ -142,14 +142,16 @@ struct InternalExtractorEdge
...
 {
     using value_type = InternalExtractorEdge;
     bool operator()(const InternalExtractorEdge &lhs, const InternalExtractorEdge &rhs) const
     {
         return (lhs.result.source < rhs.result.source) ||
                ((lhs.result.source == rhs.result.source) &&
-                (lhs.result.target < rhs.result.target));
+                (lhs.result.target < rhs.result.target ||
+                 (lhs.result.target == rhs.result.target &&
+                  lhs.result.name_id > rhs.result.name_id)));
```
Comment (MoKob) [inline include/extractor/internal_extractor_edge.hpp]: the parsing happens asynchronously, doesn't it? So the edges arrive in a different order which should also result in name_ids being assigned randomly? Or am I missing something? NameID should not be possible to use here?

File: features/testbot/basic.feature [hunk 4]
```diff
@@ -251,3 +251,34 @@ Feature: Basic Routing
...
+            | nodes | highway | name |
+            | ab    | primary | 1    |
+            | ab    | primary |      |
+            | ab    | primary | 2    |
+            | bc    | primary | 3    |
+
+        When I route I should get
+            | from | to | route |
+            | a    | c  | 2,3,3 |
+            | c    | a  | 3,2,2 |
```
Comment (MoKob) [inline features/testbot/basic.feature]: Could fail randomly, due to parallel parsing. `ab` are all identical, so they are differentiated on by the name_id which is random due to https://github.com/Project-OSRM/osrm-backend/blob/0e4061f85816b5499896022eb151a1402ae3aad0/src/extractor/extractor.cpp#L212 and its interaction with https://github.com/Project-OSRM/osrm-backend/blob/0e4061f85816b5499896022eb151a1402ae3aad0/src/extractor/extractor.cpp#L160

File: src/extractor/extraction_containers.cpp (hunk 1 above)
Comment (MoKob) [inline src/extractor/extraction_containers.cpp]: this value is never read? And the IDX is used to choose only the minimal edge. All others are invalidated. Does this even affect anything?

File: include/extractor/internal_extractor_edge.hpp (hunk 3 above)
Comment (oxidase) [inline include/extractor/internal_extractor_edge.hpp]: Yes, it fixes only the original issue with an empty name and non-empty one.

Comment (MoKob) [inline include/extractor/internal_extractor_edge.hpp]: Well... fixes. It randomly chooses any of the non-empty names over an empty name. In that case, your test-case is still wrong, since it offers multiple possible names (50:50 chance)

File: features/testbot/basic.feature (hunk 4 above)
Comment (oxidase) [inline features/testbot/basic.feature]: Frankly, i have no idea what to use for stable ordering of names. Would it make sense to take a lexicographically  smallest non-empty string?

Comment (MoKob) [inline features/testbot/basic.feature]: I think that might be an option. Otherwise we end up with just another undefined behaviour.

File: src/extractor/extraction_containers.cpp (hunk 1 above)
Comment (oxidase) [inline src/extractor/extraction_containers.cpp]: Yes, without the change min_forward_weight is always std::numeric_limits<int>::max()
and "all_edges_list[i].result.weight < min_forward_weight" is always true

Comment (MoKob) [inline src/extractor/extraction_containers.cpp]: Ah yes. You are right. Loops and no coffee don't seem to mix.

Comment (MoKob) [inline src/extractor/extraction_containers.cpp]: Can we add a test-case with multiple parallel speed-edges and empty name and check for speed?

File: features/testbot/basic.feature (hunk 4 above)
Comment (oxidase) [inline features/testbot/basic.feature]: Ok, i will update PR with as suggested

**Comment by oxidase** [comment]
Added a lexicographical compare for names to disambiguate edges. Empty strings are sorted as highest values. The price is additional memory used for "STXXLNameOffsets name_offsets".
I also changed data type to "unsigned char" because  the utf8 comparison with  char code units will sort strings incorrectly   in order Greek, Cyrillic, Latin.

File: src/extractor/extraction_containers.cpp [hunk 5]
```diff
@@ -53,6 +112,12 @@ ExtractionContainers::ExtractionContainers()
     name_lengths.push_back(0);
     name_lengths.push_back(0);
 
+    // Insert three zero offsets for empty strings and the total length
+    name_offsets.push_back(0);
+    name_offsets.push_back(0);
+    name_offsets.push_back(0);
+    name_offsets.push_back(0);
+
```
Comment (danpat) [inline src/extractor/extraction_containers.cpp]: Looks like there's one extra here?

Comment (oxidase) [inline src/extractor/extraction_containers.cpp]: No, the last offset is always the total length of the character data array and at the moment it is zero. Offsets are added after pushing data into the data  array.

File: src/extractor/extraction_containers.cpp (hunk 1 above)
Comment (oxidase) [inline src/extractor/extraction_containers.cpp]: Ok, I have added empty names to the test

File: src/extractor/extraction_containers.cpp [hunk 6]
```diff
@@ -53,6 +112,12 @@ ExtractionContainers::ExtractionContainers()
     name_lengths.push_back(0);
     name_lengths.push_back(0);
 
+    // Insert three zero offsets for empty strings and the total length
```
Comment (MoKob) [inline src/extractor/extraction_containers.cpp]: this comment needs to be adjusted, if we push four zeros. It also requires an explanation of what the different indices are supposed to be.

File: src/extractor/extraction_containers.cpp [hunk 7]
```diff
@@ -35,6 +37,63 @@ struct OSMNodeIDSTXXLLess
...
+            return lhs.result.target < rhs.result.target;
+
+        if (lhs.result.name_id == 0)
+            return false;
+
+        if (rhs.result.name_id == 0)
+            return true;
+
+        BOOST_ASSERT(!name_offsets.empty() && name_offsets.back() == name_data.size());
+        const oe::ExtractionContainers::STXXLNameCharData::const_iterator data = name_data.begin();
```
Comment (oxidase) [inline src/extractor/extraction_containers.cpp]: @MoKob I made the original comparison based on name ids due to performance and efforts, but it is not correct anyway, because if there more than one non-empty names for one road the scheduler will randomize names and it will be a problem for the graph compressor and guidance to decide that it is the same road.

The statistic must be checked for such cases. For Jakarta case it is below 0.1% when it gets to names comparison, so performance impact is not so dramatic.

Comment (TheMarex) [inline src/extractor/extraction_containers.cpp]: We can run some statistics on this. My feeling is that branch is almost never hit. This will only be used if there are multi-edges in OSM which is quite rare. Having consistent tie breaking is definitely desirable.

Comment (MoKob) [inline src/extractor/extraction_containers.cpp]: Well. We can always risk it... Its just a chance of a potential performance bottle-neck. Maybe add some commentary around this to make clear that this could impact extract performance.

The problem why I insisted on having clear distinction is the parallel-edges tests. Because they could fail randomly without this destinction. So if we have tests that have this randomness-possibility, we need to get rid of randomness.

**Comment by oxidase** [comment]
I have updated PR to avoid additional memory usage. Bad place now is a conversion from offsets to lengths for RangeTable in WriteCharData. Also lexicographical order is not  good for resolving street names ambiguity, and can result in some extremely rare strange street name changes.
For example, if edges have ambiguous names (A,C), (B,C), (D,C), (E,C) than names will be A,B,C,C.

I have also checked frequency of string comparisons: for oberbayern only 4 for 1466927 edges
```
1: Fischer-von-Erlach-Straße Lohensteinstraße
2: Lohensteinstraße Fischer-von-Erlach-Straße
3: Lohensteinstraße Fischer-von-Erlach-Straße
4: Maximilianstraße Römerkesselstraße (St 2055)
```

For Germany 600 comparisons for 24882388 edges
[strings.txt](https://github.com/Project-OSRM/osrm-backend/files/355503/strings.txt)

I don't understand at the moment why so many comparisons  for some street names like "St. Roman", but for me it looks like data problems.

**Comment by TheMarex** [comment]
> upd: Added avoiding special node ids, so only 35 string comparisons left for 24882388 edges in Germany

That looks good to me. I don't think this will be noticable.

W.r.t the quality of the tie breaking: While a consitent naming for a street is desirable, I don't think these edge case warrants a more advanced technique.
